docs/DeveloperGuide.md

```diff
@@ -219,14 +220,39 @@ This section outlines the interactions between components of the protocol gatewa
+Once `MqttIotHubAdapter` has opened a receive link to IoT hub, it can receive a message at any time. When message is received:
 
 1. `MqttIotHubAdapter` calls the `ITopicNameRouter.TryMapRouteToTopicName` method. It prepares a topic name that will be used for the PUBLISH packet to the device.
 2. `MqttIotHubAdapter` then matches the topic name against the list of subscriptions for the device. If the topic name does not match any subscription filter from the list, `MqttIotHubAdapter` rejects the message.
 3. Based on the matched subscription and requested QoS on the message, `MqttIotHubAdapter` decides which QoS to use for the PUBLISH packet. 
-4. For QoS=0 or QoS=1 messages, `MqttIotHubAdapter` will send a PUBLISH packet through the pipeline (see steps 13-16 above for CONNECT processing). The message exchange for QoS=0 is completed with this step; processing For QoS=1 messages continues with the next step. Please note that the processing for QoS=2 messages is descibed as a separate scenario following this one.
-5. `MqttIotHubAdapter` enqueues an entry with the message information (packet identifier) in an ACK-pending queue. The ACK-pending queue is used for tracking messages in-flight that await a PUBACK response for the device.
-6. Eventually (assuming normal operation), PUBACK for the PUBLISH packet is received from the device and gets through the pipeline to `MqttIotHubAdapter`
-7. `MqttIotHubAdapter` peeks at the top message from the ACK-pending queue, and if it does not match the received PUBACK packet then `MqttIotHubAdapter` will stop processing (in this case skip following steps).
-8. `MqttIotHubAdapter` dequeues the message from the ACK-pending queue and completes the message on IoT Hub.
+4. If QoS=0,
+    - `MqttIotHubAdapter` sends a PUBLISH packet through the pipeline (see steps 13-16 above for CONNECT processing)
+    - At the same time `MqttIotHubAdater` completes the message on IoT hub.
+5. If QoS=1,
+    - `MqttIotHubAdapter` posts a PUBLISH packet to `RequestAckPairProcessor` responsible for handling PUBLISH-PUBACK communication.
+    - `RequestAckPairProcessor` sends PUBLISH packet through the pipeline (see steps 13-16 above for CONNECT processing).
+    - at the same time `RequestAckPairProcessor` enqueues an entry with the message state (packet identifier) in its internal ACK-pending queue. The ACK-pending queue is used for tracking messages in-flight that await on acknowledgement message to arrive from the device.
+    - Eventually (assuming normal operation), PUBACK packet corresponding to the PUBLISH packet is received from the device and gets through the pipeline to `MqttIotHubAdapter`.
+    - `MqttIotHubAdapter` dispatches PUBACK packet to `RequestAckPairProcessor`.
+    - `RequestAckPairProcessor` peeks at the top message from its ACK-pending queue, and if it does not match the received PUBACK packet, PUBACK packet gets discarded and processing stops (following steps are skipped).
```

Comment:
Should be "...packet, the PUBACK packet gets discarded and processing stops (the following steps are skipped)."